### DISCUSSION TITLE
Unified architecture for defining custom styles

### DISCUSSION THREAD

**Discussion by ianstormtaylor** [chakra-ui/chakra-ui#3328] [Ideas]
Hey, thanks for Chakra, I've been using it to prototype a few different projects and it makes getting started quickly very easy. I've noticed though that customizing Chakra does not feel so easy, and I think it's partly because there's a lack of overall API design, so I wanted to open an issue to write up the inconsistencies/problems I saw...

I think the big issue I see is that there are too many ways of solving the same problem in Chakra, which generally leads to confusion as to when to use each. But before getting into that, even just looking at the naming system it feels inconsistent:

- There are "text styles" in `theme.textStyle` and "layer styles" in `theme.layerStyle`, but for some reason "global styles" live in `theme.styles.global`.

- There are "size variants" that are given special treatment in `Component.sizes` but "visual variants" are all grouped into `Component.variants` and only one can be active at a time.

- This all gets even more complex when trying to use `parts` to update the styles of groups of components all at once.

I feel like there are too many different features at odds here. And it makes me wonder if things could be simplified to using the "variants" concept more across the entire system. For example, I really like how [Stitches handles variants](https://stitches.dev/docs/variants) where it is just a mapping that you have full control over. Which would allow for things like...

```tsx
<Button size="sm" scheme="red" fill="solid">...</Button>
```

…and it could all be customized in the same consistent way.

Similarly, the "text styles" and "layer styles" feel weird as these separate concepts that are special-cased for global styles. It feels like something that could be implemented as simple components instead. Or, if they really need to exist as global groups of styles, why not allow for "global variants" which could solve the exact same use case:

```ts
const theme = {
  ...
  variants: {
    textStyle: {
      caption: { ... },
      heading: { ... }
    }
  }
}
```

```tsx
<Box textStyle="caption">...</Box>
```

I think the existing APIs could benefit from some consolidation so that there are fewer concepts to learn and ways of customizing the same thing.

**Comment by ianstormtaylor** [comment]
Still an issue I think.

**Comment by segunadebayo** [comment]
Thanks for the ideas mentioned here!

Converting this to a discussion as I don't think it's an issue per se.

**Comment by testerez** [comment]
Also, as new Chakra user, I was really surprised to realize that the textStyles I defined was not applied correctly to `Heading` or `FormLabel`. Looking at the docs and typescript definitions, I assumed it would just work . I don't really see the point of having those reusable styles if I can't use them on all base components.
I would much prefer something that looks more like regular CSS classes. Something that allows me to give a name to a set of styles that I can use on any component.
I'm not sure variants would be the best candidate for that since it's really specific to each components. What about using layered styles and allowing the user to define many, like we can do with CSS classes. `textStyle` could then just be deprecated.

Example:
```ts
function Example() {
  return <Box layerStyle={['selected', 'bigTitle']}>This is a box</Box>
}
```

Or maybe I miss-understood the way to use textStyles. If that's the case, please correct me.

**Reply by shayded-exe** [reply]
After using Chakra for several months now, this is one of my biggest issues with it. It's really confusing figuring out how the best way to structure everything.

I still haven't been able to find a good example of how to structure your styles in a medium to large app.

Creating a whole theme for components is laborious, so I usually end up with a bunch of style props on everything.

**Reply by herrethan** [reply]
Sadly it feels like this issue has sprawled further into disarray in v3. Using Recipes we have even more ways to configure themeable variants:

```js
defineRecipe({
  variants: {
    myVariant: { true: { color: 'green.500' } },
    // or
    variant: { myVariant: { color: 'green.500' } },
    // or
    visual: { myVisual: { color: 'green.500' } },
})
```

all of which seem to kind of work one way or another (these were all taken from various discussions). 

It's not clear how the built-in Chakra `variant` prop should play with the new `visual` prop recipe pattern. Should we be overriding `variant`? Or should we only be creating `visual`, `size` etc properties when creating component recipes? 

I've loved Chakra for years now but this v3 migration is truly a headache.
